src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1503,17 +1509,20 @@ _mongoc_topology_push_server_session (mongoc_topology_t *topology,
 {
    int64_t timeout;
    mongoc_server_session_t *ss;
+   bool loadbalanced;
 
    ENTRY;
 
    bson_mutex_lock (&topology->mutex);
 
    timeout = topology->description.session_timeout_minutes;
+   loadbalanced = topology->description.type == MONGOC_TOPOLOGY_LOAD_BALANCED;
 
    /* start at back of queue and reap timed-out sessions */
    while (topology->session_pool && topology->session_pool->prev) {
       ss = topology->session_pool->prev;
-      if (_mongoc_server_session_timed_out (ss, timeout)) {
+      /* Sessions do not expire when the topology type is load balanced. */
+      if (!loadbalanced && _mongoc_server_session_timed_out (ss, timeout)) {
```

Comment:
Why did you combine the `if` condition here but not in `_mongoc_topology_pop_server_session`?

In `_mongoc_topology_pop_server_session`, the original logic to kill the session or `break` _only_ happens when `loadbalanced` is false. When `loadbalanced` is true you just end up clearing the session pool locally.

Here, sessions never get killed if `loadbalanced` is true, but you're also leaving the session pool as-is and _not_ clearing it.